### PR TITLE
Implement CRD version conversions

### PR DESCRIPTION
**Pull request by LukeShu**
Diffstat, excluding `vendor/`:

    92 files changed, 20958 insertions(+), 1222 deletions(-)

## Description

- The first 9 commits are adding the definitions of the v2 CRDs back in, and dealing with everything that comes with that.
- The next 14 commits are cleaning up the CRDs. Improving the tests, fixing mistakes in the v2 schema, that sort of thing.
- The final 3 commits are actually implementing conversions:
  1. Add `conversion-gen` (in its own commit because of all the noise in `vendor/`)[1]
  2. Get `conversion-gen` to run.
  3. Fill in all of the conversion stubs, adding tests for the conversions. The v2 types gained a bunch of `v3*` fields so that they can store things that are new to v3alpha1. The v3alpha1 types gained a bunch of `v2*` fields for storing how something was written in v2 for all the places where there v2 had multiple ways to write something but v3alpha1 now only has one way to write it.

[1]: Also, I had to make some bug-fixes and an enhancement to `converter-gen`. View those changes at https://github.com/datawire/code-generator/compare/v0.20.2...pinned-v0.20.5-rc.0.0.20211127183116-16d402be64a9 or clone the code-generator repo and look at that commit range locally.

This does NOT implement the ambassador_id mangling to cause 1.x to not see v3alpha1 resources. As discussed on Slack, I'm not sure that's the right way to go. And even if it is, I want to get this unit of work in first.

## Related Issues
List related issues.

## Testing

This is very near to 100% test coverage: The only non-generated file that doesn't have 100% coverage is `handwritten.conversion.go`, and the lines not covered are
 - a bunch of `if err != nil { return err }` checks
 - a couple of `if out.SomeMap == nil { out.SomeMap = make(MapType) }` checks
 - A line handling `DeprecatedSelector` in the v3alpha1→v2 conversion; because it's deprecated, it *intentionally* won't losslessly round-trip back to v3alpha1.

## Checklist

 - [ ] I made sure to update `CHANGELOG.md`. - I'll leave that for (Aidan's) webhook PR
   
   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations
 
 - [x] This is unlikely to impact how Ambassador performs at scale.
 
   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability
 
 - [x] My change is adequately tested. - near 100% test coverage
 
   Remember when considering testing:
    + LEGACY MODE TESTS DO NOT RUN FOR EVERY PR. If your change is affected by legacy mode, you need
      to run legacy-mode tests manually (set `AMBASSADOR_LEGACY_MODE=true` and run the tests).
      (This will be fixed soon.)
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - not in DEVELOPING.md, but comments in the source files

### CODE REVIEW

File: pkg/api/getambassador.io/v2/handwritten.conversion.go [hunk 1]
```diff
@@ -9,26 +9,246 @@
...
+		}
+
+		// 2. outSvc
+		//
+		//                  | explicitScheme="https://" | explicitScheme="http://" |
+		//  | originateTLS  | inSvc                     | force-https://           |
+		//  | !originateTLS | not possible              | inSvc                    |
+		//
+		// Because an https:// scheme forces originateTLS=true, the bottom-right corner
+		// isn't possible
```
Comment (kflynn) [inline pkg/api/getambassador.io/v2/handwritten.conversion.go:81]: It's the bottom left that's not possible, right?

File: pkg/api/getambassador.io/v2/handwritten.conversion.go [hunk 2]
```diff
@@ -9,26 +9,246 @@
...
+		var outExplicitScheme *string
+		if (strings.ToLower(explicitScheme) == "https://") == originateTLS {
```
Comment (kflynn) [inline pkg/api/getambassador.io/v2/handwritten.conversion.go:83]: This is really easy to misread, until you look past it and read the "cover the diagonal" comment. Just sayin'. 🙂

File: pkg/api/getambassador.io/v2/handwritten.conversion.go [hunk 3]
```diff
@@ -9,26 +9,246 @@
...
+		originateTLS = true
+	case strings.HasPrefix(strings.ToLower(inSvc), "http://"):
+		bareSvc = inSvc[len("http://"):]
+	default:
+		bareSvc = inSvc
+	}
+	if inTLS != "" {
+		originateTLS = true
+	}
+	// OK, now re-serialize that in v3alpha1 syntax.
```
Comment (kflynn) [inline pkg/api/getambassador.io/v2/handwritten.conversion.go:130]: v3alpha1 syntax, or v2?

File: pkg/api/getambassador.io/v3alpha1/common.go [hunk 4]
```diff
@@ -25,6 +25,42 @@ import (
...
+// names.  But it should not affect any end-user observable behavior.
+type V2ExplicitTLS struct {
+	// TLS controls whether and how to represent the "tls" field when
+	// its value could be implied by the "service" field.
+	//
+	//  | Value        | Representation                        | Meaning of representation          |
+	//  |--------------+---------------------------------------+------------------------------------|
+	//  | ""           | omit the field                        | defer to service (no TLSContext)   |
+	//  | "null"       | store an explicit "null" in the field | defer to service (no TLSContext)   |
+	//  | "string"     | store an empty string in the field    | defer to service (no TLSContext)   |
```
Comment (kflynn) [inline pkg/api/getambassador.io/v3alpha1/common.go:41]: I thought `string` was a _non_-empty string. It'd be nice to maybe have this be `empty` or some such.

**Review by kflynn — APPROVED** [review]
Hit it! My only comments are things that aren't worth messing with until the next time you need to be messing with things anyway.

File: pkg/api/getambassador.io/v2/handwritten.conversion.go (hunk 1 above)
Comment (LukeShu) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:81]: Oops!

File: pkg/api/getambassador.io/v2/handwritten.conversion.go (hunk 2 above)
Comment (LukeShu) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:83]: It's also really easy to mis-think. IDK what to do about that.

Comment (LukeShu) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:83]: By the end I was just mashing on the keyboard trying to get the code-coverage and unit tests happy. "What do you mean the `else` case isn't getting covered! The tests are extensive! Ohhh... the `else` case is impossible... how does that make sense?"

File: pkg/api/getambassador.io/v2/handwritten.conversion.go (hunk 3 above)
Comment (LukeShu) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:130]: Oops!

File: pkg/api/getambassador.io/v3alpha1/common.go (hunk 4 above)
Comment (LukeShu) [reply pkg/api/getambassador.io/v3alpha1/common.go:41]: A _non_-empty string means that there's a TLSContext, and the V2ExplicitTLS.TLS field is ignored.  This field is mostly "how to represent an 'empty' value" (except for `bool:true`, which is special).

Comment (LukeShu) [reply pkg/api/getambassador.io/v3alpha1/common.go:41]: What about
```
	// TLS controls whether and how to represent the "tls" field when
	// its value could be implied by the "service" field.  In v2, there
	// were a lot of different ways to spell an "empty" value, and this
	// field specifies which way to spell it (if the value will indeed
	// be empty).
	//
	//  | Value        | Representation                        | Meaning of representation          |
	//  |--------------+---------------------------------------+------------------------------------|
	//  | ""           | omit the field                        | defer to service (no TLSContext)   |
	//  | "null"       | store an explicit "null" in the field | defer to service (no TLSContext)   |
	//  | "string"     | store an empty string in the field    | defer to service (no TLSContext)   |
	//  | "bool:false" | store a Boolean "false" in the field  | defer to service (no TLSContext)   |
	//  | "bool:true"  | store a Boolean "true" in the field   | originate TLS (no TLSContext)      |
```

Comment (kflynn) [reply pkg/api/getambassador.io/v3alpha1/common.go:41]: Yeah, that works.

File: pkg/api/getambassador.io/v2/handwritten.conversion.go (hunk 2 above)
Comment (LukeShu) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:83]: what about
```go
		schemeIsHTTPS := strings.ToLower(explicitScheme) == "https://"
		if schemeIsHTTPS == originateTLS {
			// cover the diagonal
			*outSvc = inSvc
		} else {
			// cover the top-right
```

Comment (kflynn) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:83]: 👍

**Comment by LukeShu** [comment]
v2: Made changes per Flynn's feedback
```patch
diff --git a/pkg/api/getambassador.io/v2/handwritten.conversion.go b/pkg/api/getambassador.io/v2/handwritten.conversion.go
index cb0a12cc1..1e94c6238 100644
--- a/pkg/api/getambassador.io/v2/handwritten.conversion.go
+++ b/pkg/api/getambassador.io/v2/handwritten.conversion.go
@@ -77,10 +77,11 @@ func convert_v2_TLS_To_v3alpha1_TLS(
 		//  | originateTLS  | inSvc                     | force-https://           |
 		//  | !originateTLS | not possible              | inSvc                    |
 		//
-		// Because an https:// scheme forces originateTLS=true, the bottom-right corner
-		// isn't possible
+		// Because an https:// scheme forces originateTLS=true, the bottom-left corner isn't
+		// possible.
 		var outExplicitScheme *string
-		if (strings.ToLower(explicitScheme) == "https://") == originateTLS {
+		schemeIsHTTPS := strings.ToLower(explicitScheme) == "https://"
+		if schemeIsHTTPS == originateTLS {
 			// cover the diagonal
 			*outSvc = inSvc
 		} else {
@@ -88,6 +89,7 @@ func convert_v2_TLS_To_v3alpha1_TLS(
 			outExplicitScheme = &explicitScheme
 			*outSvc = "https://" + bareSvc
 		}
+
 		// 3. outExplicit
 		if outExplicitTLS == "" && outExplicitScheme == nil {
 			*outExplicit = nil
@@ -127,7 +129,7 @@ func convert_v3alpha1_TLS_To_v2_TLS(
 	if inTLS != "" {
 		originateTLS = true
 	}
-	// OK, now re-serialize that in v3alpha1 syntax.
+	// OK, now re-serialize that in v2 syntax.
 	tlsIsTruthy := false
 	if inTLS != "" {
 		*outTLS = &BoolOrString{
diff --git a/pkg/api/getambassador.io/v3alpha1/common.go b/pkg/api/getambassador.io/v3alpha1/common.go
index 0e151ea3a..b1102937e 100644
--- a/pkg/api/getambassador.io/v3alpha1/common.go
+++ b/pkg/api/getambassador.io/v3alpha1/common.go
@@ -32,7 +32,10 @@ import (
 // names.  But it should not affect any end-user observable behavior.
 type V2ExplicitTLS struct {
 	// TLS controls whether and how to represent the "tls" field when
-	// its value could be implied by the "service" field.
+	// its value could be implied by the "service" field.  In v2, there
+	// were a lot of different ways to spell an "empty" value, and this
+	// field specifies which way to spell it (and will therefore only
+	// be used if the value will indeed be empty).
 	//
 	//  | Value        | Representation                        | Meaning of representation          |
 	//  |--------------+---------------------------------------+------------------------------------|
```

**Review by kflynn — APPROVED** [review]
Hit it! and thanks. 🙂

**Pull request merged**